light-client/relayer/src/relayer.rs

```diff
@@ -0,0 +1,286 @@
+/// and forward them to a "sender", together with proofs of the block's
+/// validity, when it finds any.
+pub struct RelayerThread<S, V>
+where
+    S: Sender + Send + Sync + 'static,
+    V: Verifier + Send + Sync + 'static,
+{
+    _config: Config,
+    next_block_index: BlockIndex,
+    ledger_db: LedgerDB,
+    watcher_db: WatcherDB,
+    sender: S,
+    verifier: V,
+    stop_requested: Arc<AtomicBool>,
+    shared_state: Arc<Mutex<RelayerSharedState>>,
+    logger: Logger,
+}
+
+impl<S, V> RelayerThread<S, V>
+where
+    S: Sender + Send + Sync + 'static,
+    V: Verifier + Send + Sync + 'static,
+{
+    /// Poll for new data every 10 ms
+    const POLLING_FREQUENCY: Duration = Duration::from_millis(10);
+    /// How frequently to retry if an error occurs.
+    const ERROR_RETRY_FREQUENCY: Duration = Duration::from_millis(1000);
+
+    /// Entrypoint for relayer loop.
+    pub fn start(
+        config: Config,
+        ledger_db: LedgerDB,
+        watcher_db: WatcherDB,
+        sender: S,
+        verifier: V,
+        stop_requested: Arc<AtomicBool>,
+        shared_state: Arc<Mutex<RelayerSharedState>>,
+        logger: Logger,
+    ) {
+        let next_block_index = config.start_block_index;
+        let thread = Self {
+            _config: config,
+            next_block_index,
+            ledger_db,
+            watcher_db,
+            sender,
+            verifier,
+            stop_requested,
+            shared_state,
+            logger,
+        };
+        thread.run();
+    }
+
+    pub fn run(mut self) {
+        // Poll ledger for data to relay
+        log::info!(self.logger, "Relayer thread started.");
+        loop {
+            self.shared_state().current_block_index = self.next_block_index;
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Relayer thread stop requested.");
+                break;
+            }
+            match self.ledger_db.get_block_data(self.next_block_index) {
+                Err(LedgerError::NotFound) => std::thread::sleep(Self::POLLING_FREQUENCY),
+                Err(e) => {
+                    log::error!(
+                        self.logger,
+                        "Unexpected error when checking for block data {}: {:?}",
+                        self.next_block_index,
+                        e
+                    );
+                    std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                }
+                Ok(block_data) => {
+                    if let Err(err) = self.process_block(&block_data) {
+                        log::error!(
+                            self.logger,
+                            "When processing block {}: {:?}",
+                            self.next_block_index,
+                            err
+                        );
+                        std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    } else {
+                        self.next_block_index += 1;
```

Comment:
Should update the [counters](https://github.com/mobilecoinfoundation/mobilecoin/blob/2c289ec37fa353631490b614c72208ea6b7eaa1d/light-client/relayer/src/counters.rs)